Add support for Kogan LX10 Vacuum

YodaDaCoda opened the pull request:
Mappings are incomplete. I'm unsure how to correctly enter all the values, and some are unknown at this time. Any assistance with this would be much appreciated.

It's not clear to me how to add the device to HA - I thought creating the yaml file would be enough, but I still get the message "Sorry, there is no support for this device." If I can get past this, I'll be able to make much more progress.

When complete, fixes #723

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+  entity: vacuum
+  dps:
+    - id: 101
+      name: clean_switch
```
Rename to `power`

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      name: clean_switch
+      type: boolean
+    - id: 102
+      name: pause_switch
```
Rename to `activate`

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      name: charge_switch
+      type: boolean
+    - id: 104
+      name: clean_mode
```
Rename to `command`

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: pose
+          value: pose
+        - dps_val: zone
+          value: zone
```
Maybe change one of these values to `clean_spot`, which is recognised as special by some of the vacuum cards to give you an icon on the card instead of needing to select in more info.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: zone
+          value: zone
+        - dps_val: backcharge
+          value: backcharge
```
Change this value to `return_to_base`, as that is the other specially recognised command.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: pause
+          value: pause
+        - dps_val: chargring
+          value: chargring
```
Fix the spelling in the value (even if the incorrect spelling is correct for the dps_val used in the protocol) - `charging` is a special recognised status that will affect the high level state of the entity.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: chargring
+          value: chargring
+        - dps_val: tocharge
+          value: tocharge
```
Change this value to `returning_to_base` to have it recognised by HA as a special status that will affect the state.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: tocharge
+          value: tocharge
+        - dps_val: fullcharge
+          value: fullcharge
```
Change this value to `charged` to have it affect the state correctly.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      name: pause_switch
+      type: boolean
+    - id: 103
+      name: charge_switch
```
Split this out into a separate switch entity under secondary_entities.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+    - id: 106
+      name: battery
+      type: integer
+    - id: 107
```
Split this out into a separate sensor entity under secondary_entities.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      name: cur_clean_time
+      type: integer
+    - id: 108
+      name: cur_clean_area
```
Split this out into a separate sensor entity under secondary entities.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      name: cur_clean_area
+      type: integer
+    - id: 109
+      name: fan_mode
```
Rename to fan_speed.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: auto
+          value: auto
+    - id: 110
+      name: water_mode
```
Split this out to a separate select entity under secondary_entities, as the vacuum platform does not support mopping despite repeated requests to the HA developers.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: strong
+          value: strong
+    - id: 111
+      name: remote_ctrl
```
Rename to `direction_control`

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: stop
+          value: stop
+    - id: 112
+      name: seek_robot
```
Rename to `locate`

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      optional: true
+      type: boolean
+    - id: 113
+      name: disturb_switch_1
```
Split out into seperate switch under secondary entities.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      name: disturb_switch_1
+      type: boolean
+    - id: 114
+      name: volume
```
Split out into separate number or sensor entity under secondary entities, depending on whether this is a control for how quickly to spread water during mopping, or just feedback on how much is left in the tank.

YodaDaCoda replied:
the 'volume' DP is actually the speaker volume

make-all replied:
So I think it can be a number entity then.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        min: 0
+        max: 10
+    - id: 115
+      name: material_reset
```
Probably split this out into separate buttons for each value.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: resetfilter
+          value: resetfilter
+    - id: 116
+      name: total_clean_time
```
Split this out into a separate sensor

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      type: integer
+      unit: min
+    - id: 117
+      name: total_clean_area
```
Split this out into a seperate sensor

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      type: integer
+      unit: m²
+    - id: 118
+      name: total_clean_count
```
Split this out into a separate sensor

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      name: total_clean_count
+      type: integer
+    - id: 119
+      name: side_brush_time
```
Split this out into a separate sensor

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      type: integer
+      unit: s
+    - id: 120
+      name: main_brush_time
```
Split this out into a separate sensor

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      type: integer
+      unit: s
+    - id: 121
+      name: filter_time
```
Split this out into a separate sensor

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+    - id: 122
+      name: robot_fault
+      # TODO: correctly parse the bitfield
+      type: boolean # should result in non-zero showing as true
```
Rename this to `error` to have it affect the status of the vacuum.  Type must match the underlying type, so integer or bitfield, not boolean.
If you identify the individual bits, you can create additional binary_sensors for specific errors or groups of errors under secondary_entities, and/or map the values to meaningful strings for the error attribute here.

YodaDaCoda replied:
The mappings here are taken from another integration - I have no idea how the numbers map to actual fault statuses.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: '20'
+          value: '20'
+    - id: 129
+      name: reset_map
```
Maybe split into a separate button for completeness, but the map is cloud based and won't be available, so not really useful

make-all commented:
If it is still not recognised after renaming all the dps to match vacuum entity requirements, then recheck the dps that are coming back in the log, there may be some more that need to be marked optional in the config, as they are not always being reported by the device.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+        - dps_val: backcharge
+          value: backcharge
+    - id: 105
+      name: robot_state
```
Rename to `status`

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,259 @@
+      mapping:
+        - dps_val: idle
+          value: idle
+        - dps_val: pointing\areaing
```
I'm not sure if backslash is going to get special treatment here, better avoid it for the value, and for dps_val, do some testing - it may need to be doubled to escape it, and/or the string may need double quotes.

YodaDaCoda replied:
Without quotes, the backslash appears to be interpreted correctly. When adding double quotes, it would need to be double escaped.

YodaDaCoda commented:
@make-all I'm fairly sure I've addressed all the review comments. The device still isn't detected (Device matches None with quality of 0%). Could moving DP 103 into secondary_entities have an effect there?

make-all commented:
> @make-all I'm fairly sure I've addressed all the review comments. The device still isn't detected (Device matches None with quality of 0%). Could moving DP 103 into secondary_entities have an effect there?

Where it is does not affect detection, just whether it is marked as optional or not.

In the log message on #723, dp 103 is present, so maybe another one.  I haven't gone through and checked them all against the log myself, but types also need to match.

YodaDaCoda commented:
I've checked all the DPs against the logs, types and optional status all appears to be correct to me. I'll try to make time in the coming days to dig deeper into the matching function & see where it's not correlating correctly.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,332 @@
+        type: boolean
+  - entity: button
+    name: material_reset_gettime
+    category: diagnostic
```
category should be config for buttons.

make-all commented on custom_components/tuya_local/devices/kogan_lx10_vacuum.yaml:
```diff
@@ -0,0 +1,332 @@
+        unit: m
+  - entity: number
+    name: cur_clean_area
+    category: diagnostic
```
category should be config for numbers, but if the entity is a sensor, then diagnostic is correct.

make-all reviewed (commented):
Some of the comments I didn't repeat over and over (only the ones which are actual errors that will stop the config from working).

make-all reviewed (changes requested):
Some of the comments I didn't repeat over and over (only the ones which are actual errors that will stop the config from working).
